src/C++/SSLSocketConnection.cpp

```diff
@@ -182,52 +184,62 @@ bool SSLSocketConnection::processQueue()
     return false;
     
   const std::string& msg = m_sendQueue.front();
-  m_sendLength = 0;
-
-  while (m_sendLength < (int)msg.length())
+  
+  errno = 0;
+  int errCodeSSL = 0;
+  int sent = 0;
+  ERR_clear_error();
+
+  sent = SSL_write(m_ssl, msg.c_str() + m_sendLength,  msg.length() - m_sendLength);
+      
+  if (sent > 0)
   {
-    errno = 0;
-    int errCodeSSL = 0;
-    int sent = 0;
-    ERR_clear_error();
-
-    // Cannot do concurrent SSL write and read as ssl context has to be
-    // protected. Done above.
-    {
-      sent = SSL_write(m_ssl, msg.c_str() + m_sendLength,  msg.length() - m_sendLength);
-      if (sent <= 0)
-        errCodeSSL = SSL_get_error(m_ssl, sent);
-    }
+      m_sendLength += sent;
 
-    if (sent <= 0)
-    {
-      if ((errCodeSSL == SSL_ERROR_WANT_READ) ||
-          (errCodeSSL == SSL_ERROR_WANT_WRITE))
+      if (m_sendLength == msg.length())
       {
-        errno = EINTR;
-        sent = 0;
+          m_sendLength = 0;
+          m_sendQueue.pop_front();
       }
-      else
-      {
-        char errbuf[200];
+      return m_sendQueue.empty();
+  }
 
-        socket_error(errbuf, sizeof(errbuf));
+  errCodeSSL = SSL_get_error(m_ssl, sent);
 
-        m_pSession->getLog()->onEvent("SSL send error <" +
-                                      IntConvertor::convert(errCodeSSL) + "> " +
-                                      errbuf);
+  if ((errCodeSSL == SSL_ERROR_WANT_READ) ||
+      (errCodeSSL == SSL_ERROR_WANT_WRITE))
+  {
+      errno = EINTR;
+      sent = 0;
 
-        return m_sendQueue.empty();
+      if (errCodeSSL == SSL_ERROR_WANT_WRITE) {
+          return false;
       }
-    }
 
-    m_sendLength += sent;
+      if (errCodeSSL == SSL_ERROR_WANT_READ)
+      {
+          m_processQueueNeedsToReadData = true;
+    	  return true; //we may have more to write but we want to read first, so unsignal for now
+      }
   }
+  else
+  {
+      char errbuf[200];
 
-  m_sendLength = 0;
-  m_sendQueue.pop_front();
+      socket_error(errbuf, sizeof(errbuf));
 
-  return m_sendQueue.empty();
+      m_pSession->getLog()->onEvent("SSL send error <" +
+                                      IntConvertor::convert(errCodeSSL) + "> " +
+                                      errbuf);
+
+      return false;
+  }  
+}
```

Comment:
Hello, I'm using your PR with a custom build for now.
I noticed the following error with g++ 11:

```
/home/McLeodD/quickfix-ssl/src/C++/SSLSocketConnection.cpp:237:1: warning: control reaches end of non-void function [-Wreturn-type]
  237 | }

```

I think the compiler has a hard time to see that every branch return something. The 'else' should be removed.